File: ios/sdk/swift/internal/ContactMessage.swift
import CrossPL

extension ContactInternal {

/* @CrossClass */
open class ContactMessage: CrossBase {
  public enum Kind: Int {
    case Empty       = 0x00000000
    case MsgText     = 0x00000001
    case MsgAudio    = 0x00000002
    case MsgTransfer = 0x00000004
    case MsgImage    = 0x00000008
    case MsgFile     = 0x00000010
    case MsgBinary   = 0x00000020
  }
  
  public class MsgData {
    public func toString() -> String {
      let data = toData()
      return Data.ToString(from: data) ?? "nil"
    }
    public func toData() -> Data? {
      fatalError("\(#function) not implementation.")
    }
    public func fromData(data: Data?) {
      fatalError("\(#function) not implementation.")
    }
  }

  public class TextData: MsgData {
    public init(text: String?) {
      self.text = text;
      super.init()
    }
    
    public override func toData() -> Data? {
      String.ToData(from: self.text)!
    }
    
    public override func fromData(data: Data?) {
      self.text = Data.ToString(from: data)
    }
    
    public private(set) var text: String?
  }
  
  public class BinaryData: MsgData {
    public init(data: Data?) {
      self.data = data;
      super.init()
    }
    
    public override func toString() -> String {
      let encode = try! JSONEncoder().encode(self.data)
      let val = String(data: encode, encoding: .utf8)!
      return val
    }
    
    public override func toData() -> Data? {
      return self.data
    }
    
    public override func fromData(data: Data?) {
      self.data = data
    }
    
    public private(set) var data: Data?
  }

  public class FileData: MsgData, Codable {
    public init(file: URL?) {
      devId = UserInfo.GetCurrDevId()
      if(file != nil) {
        name = file?.lastPathComponent
        let attr = try? FileManager.default.attributesOfItem(atPath: file!.path)
        size = attr![FileAttributeKey.size] as? Int64
        md5 = Utils.getMD5Sum(file: file)
      }
      super.init()
    }
    
    public override func toData() -> Data? {
      let json = try? JSONEncoder().encode(self)
      return json
    }
    
    public override func fromData(data: Data?) {
      let newData = try? JSONDecoder().decode(FileData.self, from: data!)
      self.devId = newData?.devId
      self.name = newData?.name
      self.size = newData?.size
      self.md5 = newData?.md5
     }
    
    public required init(from decoder: Decoder) throws {
      let container = try decoder.container(keyedBy: CodingKeys.self)
      devId = try container.decode(String.self, forKey: .devId)
      name = try container.decode(String.self, forKey: .name)
      size = try container.decode(Int64.self, forKey: .size)
      md5 = try container.decode(String.self, forKey: .md5)
    }
    
    public func encode(to encoder: Encoder) throws {
      var container = encoder.container(keyedBy: CodingKeys.self)
      try container.encode(devId, forKey: .devId)
      try container.encode(name, forKey: .name)
      try container.encode(size, forKey: .size)
      try container.encode(md5, forKey: .md5)
    }
    
    // fix json decode and encode different issue
    public static func ConvertId(id: String) -> String? {
      guard let dataId = id.data(using: .utf8) else {
        print("FileData.ConvertId() 0 Failed to convert \(id)")
        return nil
      }
      
      guard let fileData = try? JSONDecoder().decode(FileData.self, from: dataId) else {
        print("FileData.ConvertId() 1 Failed to convert \(id)")
        return nil
      }
      
      return Data.ToString(from: fileData.toData())
    }
    
    public private(set) var devId: String?
    public private(set) var name: String?
    public private(set) var size: Int64?
    public private(set) var md5: String?
    
    enum CodingKeys : String, CodingKey {
      case devId
      case name
      case size
      case md5
      var rawValue: String {
        get {
          switch self {
            case .devId: return JsonKey.DeviceId
            case .name: return JsonKey.Name
            case .size: return JsonKey.Size
            case .md5: return JsonKey.Md5
          }
        }
      }
    }
  }

  
  public let type: Kind
  public private(set) var data: MsgData
  public let cryptoAlgorithm: String?
  public var timestamp: Int64
  
  public func syncMessageToNative() -> Int {
    let ret = syncMessageToNative(type: type.rawValue,
                                  data: data.toData()!,
                                  cryptoAlgorithm: cryptoAlgorithm,
                                  timestamp: timestamp)
    return ret;
  }


  public init(type: Kind, data: MsgData, cryptoAlgorithm: String?) {
    self.type = type
    self.data = data
    self.cryptoAlgorithm = cryptoAlgorithm ?? ""
    self.timestamp = Int64(Date().timeIntervalSince1970 * 1000)
    
    super.init(className: String(describing: ContactMessage.self))
  }

  public convenience init(text: String, cryptoAlgorithm: String?) {
    self.init(type: Kind.MsgText,
              data: TextData(text: text),
              cryptoAlgorithm: cryptoAlgorithm);
  }

  public convenience init(file: URL, cryptoAlgorithm: String?) {
    self.init(type: Kind.MsgFile,
              data: FileData(file: file),
              cryptoAlgorithm: cryptoAlgorithm);
  }
  
  public convenience init(type: Kind, data: Data, cryptoAlgorithm: String?) {
    self.init(type: type, data: MsgData(), cryptoAlgorithm: cryptoAlgorithm)
    switch (type) {
    case .MsgText:
      self.data = TextData(text: nil)
      break;
    case .MsgBinary:
      self.data = BinaryData(data: nil)
      break;
    case .MsgFile:
      self.data = FileData(file: nil)
      break;
    default:
      print("Unknown message type \(type)")
      return;
    }
    self.data.fromData(data: data)
  }
  
  /* @CrossNativeInterface */
  private func syncMessageToNative(type: Int,
                                   data: Data, cryptoAlgorithm: String?,
                                   timestamp: Int64) -> Int {
    let ret = crosspl_Proxy_ContactMessage_syncMessageToNative(nativeHandle, Int32(type), data, cryptoAlgorithm, timestamp)
    return Int(ret)
  }
}

}

File: android/sdk/src/main/java/org/elastos/sdk/elephantwallet/contact/internal/ContactMessage.java
package org.elastos.sdk.elephantwallet.contact.internal;

import android.util.Log;

import com.google.gson.Gson;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

import org.elastos.sdk.elephantwallet.contact.Utils;
import org.elastos.tools.crosspl.CrossBase;
import org.elastos.tools.crosspl.annotation.CrossClass;
import org.elastos.tools.crosspl.annotation.CrossInterface;

import java.io.File;
import java.util.Random;

@CrossClass
class ContactMessage extends CrossBase {
    public enum Type {
        Empty(0x00000000),
        MsgText(0x00000001),
        MsgAudio(0x00000002),
        MsgTransfer(0x00000004),
        MsgImage(0x00000008),
        MsgFile(0x00000010),
        MsgBinary(0x00000020);

        public static Type valueOf(int id) {
            Type[] values = Type.values();
            for(int idx = 0; idx < values.length; idx++) {
                if(values[idx].id == id) {
                    return values[idx];
                }
            }
            return null;
        }

        private Type(int id){
            this.id = id;
        }
        private int id;
    }

    public static abstract class MsgData {
        public String toString() {
            String str = Utils.GetGsonBuilder().create().toJson(this);
            return str;
        }

        public abstract byte[] toData();
        public abstract void fromData(byte[] data);
    }

    public static class TextData extends MsgData {
        public TextData() {
        }
        public TextData(String text) {
            this.text = text;
        }

        @Override
        public String toString() {
            return text;
        }
        @Override
        public byte[] toData() {
            return text.getBytes();
        }
        @Override
        public void fromData(byte[] data) {
            text = new String(data);
        }

        public String text;
    }

    public static class BinaryData extends MsgData {
        public BinaryData() {
        }
        public BinaryData(byte[] binary) {
            this.binary = binary;
        }

        @Override
        public String toString() {
            String str = Utils.GetGsonBuilder().create().toJson(binary);
            return str;
        }
        @Override
        public byte[] toData() {
            return binary;
        }
        @Override
        public void fromData(byte[] data) {
            binary = data;
        }

        public byte[] binary;
    }

    public static class FileData extends MsgData {
        public FileData() {
        }
        public FileData(File file) {
            devId = UserInfo.GetCurrDevId();
            name = file.getName();
            size = file.length();
            md5 = Utils.getMD5Sum(file);
        }

        @Override
        public String toString() {
            byte[] data = toData();
            return new String(data);
        }
        @Override
        public byte[] toData() {
            JsonObject json = new JsonObject();
            json.addProperty(JsonKey.DeviceId, this.devId);
            json.addProperty(JsonKey.Name, this.name);
            json.addProperty(JsonKey.Size, this.size);
            json.addProperty(JsonKey.Md5, this.md5);

            return json.toString().getBytes();
        }
        @Override
        public void fromData(byte[] data) {
            JsonElement element = new JsonParser().parse(new String(data));
            JsonObject json = element.getAsJsonObject();
            this.devId = json.get(JsonKey.DeviceId).getAsString();
            this.name = json.get(JsonKey.Name).getAsString();
            this.size = json.get(JsonKey.Size).getAsLong();
            this.md5 = json.get(JsonKey.Md5).getAsString();
        }

        // fix json decode and encode different issue
        public static String ConvertId(String id) {
            FileData fileData = Utils.GetGsonBuilder().create().fromJson(id, FileData.class);
            if (fileData == null) {
                Log.w(Utils.TAG, "FileData.ConvertId() 0 Failed to convert " + id);
            }

            return fileData.toString();
        }

        public String devId;
        public String name;
        public long size;
        public String md5;
    }

    public final Type type;
    public MsgData data;
    public final String cryptoAlgorithm;
    public long nanoTime;
    public long replyToNanoTime;

    public void replyTo(long toNanoTime) {
        this.replyToNanoTime = toNanoTime;
    }

    int syncMessageToNative() {
        byte[] msgData = data.toData();
        int ret = syncMessageToNative(type.id, msgData, cryptoAlgorithm, nanoTime, replyToNanoTime);
        return ret;
    }

    ContactMessage(String text, String cryptoAlgorithm) {
        this(Type.MsgText, new TextData(text), cryptoAlgorithm);
    }

    ContactMessage(byte[] binary, String cryptoAlgorithm) {
        this(Type.MsgBinary, new BinaryData(binary), cryptoAlgorithm);
    }


    ContactMessage(File file, String cryptoAlgorithm) {
        this(Type.MsgFile, new FileData(file), cryptoAlgorithm);
    }

    ContactMessage(Type type, byte[] data, String cryptoAlgorithm) {
        this(type, (MsgData) null, cryptoAlgorithm);
        if(data != null) {
            switch (type) {
                case MsgText:
                    this.data = new TextData();
                    break;
                case MsgBinary:
                    this.data = new BinaryData();
                    break;
                case MsgFile:
                    this.data = new FileData();
                    break;
                default:
                    Log.w(Utils.TAG, "Unknown Message Type: " + type);
                    break;
            }
        }
        if(data != null) {
            this.data.fromData(data);
        }
    }

    private ContactMessage(Type type, MsgData data, String cryptoAlgorithm) {
        super(ContactMessage.class.getName(), 0);

        this.type = type;
        this.data = data;
        this.cryptoAlgorithm = cryptoAlgorithm;
        this.nanoTime = System.currentTimeMillis() * 1000000 + new Random().nextInt(100000);
        this.replyToNanoTime = 0;
    }

    @CrossInterface
    private native int syncMessageToNative(int type, byte[] data, String cryptoAlgorithm,
                                           long nanoTime, long replyToNanoTime);
} // class Factory
